ymer/src/main/java/com/avanza/ymer/PersistedInstanceIdRecalculationService.java

```diff
@@ -57,21 +60,27 @@ public void recalculatePersistedInstanceId(String collectionName, int numberOfIn
 		if (!indexDropped) {
 			log.info("Step 1/3\tNo index to drop");
 		}
-		AtomicInteger count = new AtomicInteger();
-		collection.findByQuery(query(where(DOCUMENT_ROUTING_KEY).exists(true)))
-				.peek(it -> {
-					int currentCount = count.incrementAndGet();
-					if (currentCount % 10_000 == 0) {
-						log.info("Step 2/3\tUpdated persisted instance id for {} documents", currentCount);
-					}
-				})
-				.forEach(document -> {
-					Object id = document.get("_id");
-					int instanceId = getInstanceId(document.get(DOCUMENT_ROUTING_KEY), numberOfInstances);
-					collection.updateById(id, Map.of(DOCUMENT_INSTANCE_ID, instanceId));
-				});
 
-		log.info("Step 2/3\tUpdated persisted instance id for {} documents", count.get());
+		try (Stream<List<Document>> batches = collection.findByQuery(query(where(DOCUMENT_ROUTING_KEY).exists(true)), BATCH_SIZE)) {
+			AtomicInteger count = new AtomicInteger();
+			batches.forEach(batch -> {
+				List<Document> updates = batch.stream()
+						.peek(it -> {
+							int currentCount = count.incrementAndGet();
+							if (currentCount % 10_000 == 0) {
+								log.info("Step 2/3\tUpdated persisted instance id for {} documents", currentCount);
+							}
+						})
+						.map(it -> {
+							Object id = it.get("_id");
+							int instanceId = getInstanceId(it.get(DOCUMENT_ROUTING_KEY), numberOfInstances);
+							return new Document("_id", id).append(DOCUMENT_INSTANCE_ID, instanceId);
+						})
+						.collect(toList());
+				collection.updateAllPartial(updates);
```

Comment:
The commands that are issued to the mongodb server are now like this:
```
{
  "update": "collectionName",
  "updates": [
    {
      "q": { "_id": "id1" },
      "u": { "$set": { "_instanceId": 2 } }
    },
    {
      "q": { "_id": "id2" },
      "u": { "$set": { "_instanceId": 2 } }
    },
    {
      "q": { "_id": "id3" },
      "u": { "$set": { "_instanceId": 2 } }
    },
    {
      "q": { "_id": "id4" },
      "u": { "$set": { "_instanceId": 1 } }
    }
  ]
}
```

If we want to reduce the size of the commands being sent to the db server, another way of doing this would be to batch these updates per instanceid and issue commands like:
```
{
  "update": "collectionName",
  "updates": [
    {
      "q": { "_id": { "$in": [ "id1", "id2", "id3" ] } },
      "u": { "$set": { "_instanceId": 2 } }
    }
  ]
}
```

For example, the update-command in java code could be
```java
collection.updateMany(
	Filters.in("_id", ids),
	Updates.set(DOCUMENT_INSTANCE_ID, instanceId)
);
```

Reply:
Good idea! 👍